src/java/org/apache/cassandra/streaming/async/NettyStreamingMessageSender.java

```diff
@@ -93,10 +95,24 @@
     private Channel controlMessageChannel;
 
     // note: this really doesn't need to be a LBQ, just something that's thread safe
-    private Collection<ScheduledFuture<?>> channelKeepAlives = new LinkedBlockingQueue<>();
+    private final Collection<ScheduledFuture<?>> channelKeepAlives = new LinkedBlockingQueue<>();
 
     private final ThreadPoolExecutor fileTransferExecutor;
 
+    /**
+     * A {@link ThreadLocal} used by the threads in {@link #fileTransferExecutor} to stash references to constructed
+     * and connected {@link Channel}s.
+     */
+    private static final FastThreadLocal<Channel> threadLocalChannel = new FastThreadLocal<>();
+
+    /**
+     * A netty channel attribute used to indicate if a channel is currently transferring a file. This is primarily used
```

Comment:
Normally, yes, that would be completely true - that things are racey. However, since the `KeepAliveTask` executes as a task in the netty event loop (which is single-threaded for that channel), I'm guaranteed that a `KeepAliveTask` and a `FileStreamTask` will not execute concurrently in the same channel (and both try to update that channel attribute).

I will add some comments to that effect, though; that is good to have in there.